source/opt/type_manager.cpp

```diff
@@ -654,28 +646,32 @@ Type* TypeManager::RecordIfTypeDefinition(const Instruction& inst) {
       const Instruction* length_constant_inst = id_to_constant_inst_[length_id];
       assert(length_constant_inst);
 
-      // If it is a specialised constants, retrieve its SpecId.
+      // If it is a specialised constant, retrieve its SpecId.
+      // Only OpSpecConstantOp has a SpecId.
       uint32_t spec_id = 0u;
```

Comment:
Maybe this default should be `~0u`? 0 is reasonable spec id that would be used. This makes undecorated constants look the same as some decorated constants. Another option might be to use two words: one for the spec id and one to indicate if it was decorated.

Reply:
I added a boolean flag to say if it was initialized, and distinguished the case based on the flag instead.